Update Auto Detect

Chadnaut opened the pull request:
Auto detect improvements:
- Displays number of emulators found, and lists the first 3
- Detect confirm option defaults to "Yes"
- Removed confirm option when run from Emulator menu
- Detection no longer overwrites `emulator.cfg` - occurs only when `Import` confirmed

Updated import scripts:
- Updated output formatting
- Fixed `workdir`
- Fixed `executable` name on missing emu (was printing array)
- More "Additional Files" and directories
- More artwork resources and directories

Test
- This will overwrite your emu/romlist so run in sandbox
- Put files in expected locations
  - Emu at `c:\mame`
  - INI files at `c:\mame` or `c:\mame\folders` <- yes `folders` is the official recommended path!
  - Artwork at `c:\mame\snaps` or `$HOME/mame/snaps`
- Execute first-run (or config->emulators->auto_detect)

Chadnaut commented:
The user should see what was found, even if just one, to confirm expectations.
> I put Mame here. I expect you to find Mame. You found an emulator! Was it Mame?
<img width="320" height="240" alt="image" src="https://github.com/user-attachments/assets/e525fd0f-b79e-4bab-a7bf-0abe11ac3397" />

Did you put the updated `emulators/script` files in your test sandbox?
Any console errors? Nothing after "Auto-Detect"?

That reminds me, I should add a "No emulators found" message.

oomek commented:
No, all my rom paths are mapped to folders that are also location of my mame torrents. I'm sharing my 1GB upload bandwidth as much as I can.

oomek commented:
Works now. 1 emu detected. 0 roms.
Do you think you could go 1 step further and extract  `rompath` from `mame.ini` ?

Chadnaut commented:
- Updated `mame.ini` parser, now allows values with spaces (not sure how your paths were parsed previously)
- Updated formatting to match path dir-slash with OS

oomek commented:
I did not say I had roms detected previously, just that the emulator was detected.
Anyway, I've tried new scripts and I have an infinite loop. I've checked the processes and It's constantly launching and closing `mame.exe` I had to force close it after 10 minutes.

Chadnaut commented:
I meant previously *outside* of this branch.
Has the auto-detect script *ever* worked for you?
As in populated `mame.cfg` with all the correct paths, as well as successfully created a romlist.
Your `mame.ini` `rompath` has spaces that the script in master would have eaten.

Anything in the console during this infinite loop? What is the frontend displaying?

oomek commented:
Here is the log from your branch when it freezes
```
D:\EMU\ATTRACT-TEST>attractplus-console --loglevel debug
--------------------------------------------------------------------------------
Attract-Mode Plus v3.2.0 (Windows, SFML 3.0.1 +7z +Curl)
avcodec 61.19.101 / avformat 61.7.100 / swscale 8.3.100 / avutil 59.39.100 / swresample 5.3.100

NvAPI: nvapi64.dll loaded
NvAPI: Initialize: NVAPI_OK
NvAPI: CreateSession: NVAPI_OK
NvAPI: LoadSettings: NVAPI_OK
NvAPI: FindProfileByName: NVAPI_OK
NvAPI: Threaded Optimizations OFF: NVAPI_OK
NvAPI: Triple Buffering OFF: NVAPI_OK
NvAPI: Prendered Frames set to 4: NVAPI_OK
NvAPI: Maximum frames allowed set to 4: NVAPI_OK
NvAPI: GPU Maximum Performance ON: NVAPI_OK
NvAPI: Antialiasing Gamma Correction OFF: NVAPI_OK
NvAPI: Saving Nvidia profile: NVAPI_OK
NvAPI: Closing Nvidia session: NVAPI_OK

Config file not found: D:\EMU\ATTRACT-TEST\attract.cfg
Joysticks after mapping:
ID: 0 => Joy7(No Joystick)
ID: 1 => Joy7(No Joystick)
ID: 2 => Joy7(No Joystick)
ID: 3 => Joy7(No Joystick)
ID: 4 => Joy7(No Joystick)
ID: 5 => Joy7(No Joystick)
ID: 6 => Joy7(No Joystick)
ID: 7 => Joy7(No Joystick)
Created Attract-Mode Plus Window: 3442x1442 @ -1,-1 [OpenGL surface: 3442x1442 bpp=32]
Monitor's Refresh Rate: 144 Hz
[Transition] type=ShowOverlay, var=22
[Transition] type=HideOverlay, var=0
Performing emulator auto-detection
Running script: D:\EMU\ATTRACT-TEST\emulators/script/init.nut
Running script: D:\EMU\ATTRACT-TEST\modules\file.nut
Done script: D:\EMU\ATTRACT-TEST\modules\file.nut
 * atari800-5200  [Not found]
 * daphne         [Not found]
 * dolphin-emu    [Not found]
 * dosbox         [Not found]
 * fbalpha        [Not found]
 * flycast        [Not found]
 * fs-uae         [Not found]
 * fusion         [Not found]
 * gens           [Not found]
Running script: D:\EMU\ATTRACT-TEST\modules\file.nut
Done script: D:\EMU\ATTRACT-TEST\modules\file.nut
```

Chadnaut commented:
- Added more checks to help prevent detection issues
- Added more logging to help determine failure point
- XML parser now stops when all roms found (instead of continuing to parse the remaining xml)

oomek commented:
It's still launching mame in a loop
```
Attract-Mode Plus v3.2.0 (Windows, SFML 3.0.1 +7z +Curl)
avcodec 61.19.101 / avformat 61.7.100 / swscale 8.3.100 / avutil 59.39.100 / swresample 5.3.100

Config file not found: D:\EMU\ATTRACT-TEST\attract.cfg
Performing emulator auto-detection

 - Generate       atari800-5200
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/atari800-5200.cfg (Overwrite)

 - Generate       daphne
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/daphne.cfg (Overwrite)

 - Generate       dolphin-emu
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/dolphin-emu.cfg (Overwrite)

 - Generate       dosbox
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/dosbox.cfg (Overwrite)

 - Generate       fbalpha
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/fbalpha.cfg (Overwrite)

 - Generate       flycast
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/flycast.cfg (Overwrite)

 - Generate       fs-uae
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/fs-uae.cfg (Overwrite)

 - Generate       fusion
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/fusion.cfg (Overwrite)

 - Generate       gens
 - Executable     -
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/gens.cfg (Overwrite)

 - Generate       mame
 - Searching for exe...
 - Found exe      D:\EMU\mame\mame.exe
 - Running        D:\EMU\mame\mame.exe -help
 - Found version  281
 - Parsing config...
 - Running        D:\EMU\mame\mame.exe -showconfig
 - Found homepath D:\EMU\mame\
 - Found rompath  D:\Downloads\MAME ROMs (split)\
 - Found rompath  D:\Downloads\MAME BIOS\
 - Found rompath  D:\Downloads\MAME Software List ROMs (split)\
 - Found rompath  D:\EMU\MAMEDATA\ROMS\CHD\
 - Running        D:\EMU\mame\mame.exe -listmedia
 - Running        D:\EMU\mame\mame.exe -listfull
 - Running        D:\EMU\mame\mame.exe -listmedia
...
 - Running        D:\EMU\mame\mame.exe -listfull
 - Running        D:\EMU\mame\mame.exe -listmedia
 - Running        D:\EMU\mame\mame.exe -listfull
 ```

oomek commented:
Wait, is it generating list.xml for each rom? 😅

Chadnaut commented:
Most of `mame_init.nut` is still original code.
It's looking for "console systems", which it assumes are subdirs of the rom path(s).
So it's running the `SystemParser` for *every subdir*, which it appears you have a bunch of.

The previous rompath parser split on spaces.
Since I "fixed" it, and your paths have spaces, the detector is likely checking stuff it never used to.
Comment out the entire `if ( ver >= 162 )` block and see if it works without it.
In the meantime I think I have enough info to work on a fix.

oomek commented:
<img width="1442" height="1116" alt="image" src="https://github.com/user-attachments/assets/5cd5e549-008a-4f17-bd32-02c246ba1921" />
No issues after commenting out the block.

```
 - Generate       mame
 - Searching for exe...
 - Found exe      D:\EMU\mame\mame.exe
 - Running        D:\EMU\mame\mame.exe -help
 - Found version  281
 - Parsing config...
 - Running        D:\EMU\mame\mame.exe -showconfig
 - Found homepath D:\EMU\mame\
 - Found rompath  D:\Downloads\MAME ROMs (split)\
 - Found rompath  D:\Downloads\MAME BIOS\
 - Found rompath  D:\Downloads\MAME Software List ROMs (split)\
 - Found rompath  D:\EMU\MAMEDATA\ROMS\CHD\
 - Searching for extras...
 - Found extra    D:\EMU\mame\catver.ini
 - Writing templates...
 - Executable     D:\EMU\mame\mame.exe
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg (Overwrite)
 - Template       D:\EMU\ATTRACT-TEST\emulators/templates/mame-nes.cfg
```

```
Done script: D:\EMU\ATTRACT-TEST\emulators/script/init.nut
Found 1 emulator(s)
[Transition] type=ShowOverlay, var=22
[Transition] type=HideOverlay, var=0
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] executable      = D:\EMU\mame\mame.exe
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] args            = [name]
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] workdir         = D:\EMU\mame\
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] rompath         = D:\Downloads\MAME ROMs (split)\;D:\Downloads\MAME BIOS\;D:\Downloads\MAME Software List ROMs (split)\;D:\EMU\MAMEDATA\ROMS\CHD\
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] romext          = .zip;.7z
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] system          = Arcade
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] info_source     = listxml
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] import_extras   = D:\EMU\mame\catver.ini
[D:\EMU\ATTRACT-TEST\emulators/templates/mame.cfg] artwork         = snap            D:\EMU\mame\snap\
Writing emulator config to: D:\EMU\ATTRACT-TEST\emulators/mame.cfg
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] executable      = D:\EMU\mame\mame.exe
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] args            = [name]
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] workdir         = D:\EMU\mame\
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] rompath         = D:\Downloads\MAME ROMs (split)\;D:\Downloads\MAME BIOS\;D:\Downloads\MAME Software List ROMs (split)\;D:\EMU\MAMEDATA\ROMS\CHD\
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] romext          = .zip;.7z
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] system          = Arcade
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] info_source     = listxml
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] import_extras   = D:\EMU\mame\catver.ini
[D:\EMU\ATTRACT-TEST\emulators/mame.cfg] artwork         = snap            D:\EMU\mame\snap\
*** Generating Collection/Rom List: mame
 - Found 44112 files.
 - Obtaining -listxml info...96 %
 - Discarded 1505 entries based on xml info: 3dobios airlbios aleck64
  ...
 oard steeltal_board steeltal1_board steeltalp_board strtdriv_board hdrivair_board hdrivair0_board
[Import D:\EMU\mame\catver.ini] - found info for 40290 entries.
 - Removing any duplicate entries...
 + Writing 42600 entries to: D:\EMU\ATTRACT-TEST\romlists/mame.txt
[D:\EMU\ATTRACT-TEST\default-display.cfg] layout          = Basic
[D:\EMU\ATTRACT-TEST\default-display.cfg] in_cycle        = yes
[D:\EMU\ATTRACT-TEST\default-display.cfg] in_menu         = yes
[D:\EMU\ATTRACT-TEST\default-display.cfg] filter          = All
[D:\EMU\ATTRACT-TEST\default-display.cfg] filter          = Favourites
[D:\EMU\ATTRACT-TEST\default-display.cfg] rule            = Favourite equals 1
Writing config to: D:\EMU\ATTRACT-TEST\attract.cfg
 ```
